hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMOpenKeysDeleteRequest.java

```diff
@@ -155,33 +157,57 @@ private OmBucketInfo updateTableCachePerBucket(OzoneManager ozoneManager,
     OmBucketInfo omBucketInfo;
 
     try {
-      acquiredLock = omMetadataManager.getLock().acquireWriteLock(BUCKET_LOCK,
-              volumeName, bucketName);
+      acquiredLock = omMetadataManager.getLock()
+          .acquireWriteLock(BUCKET_LOCK, volumeName, bucketName);
       omBucketInfo = getBucketInfo(omMetadataManager, volumeName, bucketName);
+      final BucketLayout bucketLayout;
+      if (omBucketInfo != null) {
+        bucketLayout = omBucketInfo.getBucketLayout();
+      } else {
+        LOG.warn("Bucket {}/{} is not found", volumeName, bucketName);
```

Comment:
@kaijchen I couldn't understand this, could you please tell me the reason to handle non-existent bucket?

Reply:
The bucket might be deleted while the key remains in open key(file) table.

Actually there will be problems if the bucket is deleted and then re-created.
We are discussing this in HDDS-5867.

Reply:
Since `bucket#objectID` is used a `parentID` for the `openKey`, the deleted bucket's openKey won't be reachable. 

For ex: `buck-1`'s objectID = 512. I have writing to an open file "buck-1/file-1". Then the keyname entry in OpenFileTable will be `512/file-1`.
Say, buck-1 got deleted and re-created again. That means, the new `buck-1`'s objectID = 2000. Using this we can't traverse `512/file-1`. This will be orphan always.

Reply:
Is `parentID` set in `OmKeyInfo` of legacy keys also?

Reply:
No, parentID has been introduced in FSO. 

Say key path="/vol-1/buck-1/file-1", its corresponding `dbKeyName` representation will be:

BucketLayout.LEGACY: This uses OpenKeyTable and the dbKeyName will be "/vol-1/buck-1/file-1"
BucketLayout.FILE_SYSTEM_OPTIMIZED: This uses OpenFileTable and the dbKeyName will be "512/file-1"

Reply:
We are updating `OmBucketInfo` here because we need to update bucket quota (usedBytes and usedNamespace).

If the bucket has been deleted, the orphan keys are still reachable from `OpenKeyTable` or `OpenFileTable`.
In this case, we should not update `OmBucketInfo`.

If the bucket was re-created (before running open key cleanup service), we need a way to identify it.

Reply:
>If the bucket has been deleted, the orphan keys are still reachable from OpenKeyTable or OpenFileTable.

In case of FSO buckets, orphan keys present in `OpenFileTable` won't be reachable. Because the traversal algorithm requires objectID of the deleted bucket to identify orphan `"512/file-1"`, which is 512 and unique value. 

If a user re-creates deleted bucket with the same name `buck-1`, then its bucket objectID will be a new value say 2000. Hope you got my thoughts?

Reply:
Thanks for explaining, I got it.
But we still have to update quota if the bucket exists and is the original one.

Reply:
>But we still have to update quota if the bucket exists and is the original one.

Good point. Following are the diff cases:

*Bucket Deleted:* Then determine the FSO key using its ` parentID > 0` condition.

*Bucket re-created with same name:* 
+case-1)+ Re-created bucket with same layout - Here the key's parentID will not be matching with the newly created bucket. 
+case-2)+ Re-created bucket with diff layout - Assume I had an old FSO bucket "buck-1" and its key is "buck-1/file-1", which will be represented with the dbkey "512/file-1". Now the bucket has been re-created in OBS, its bucket ID will be 2000. 

How about will have a common logic to handle the above cases.
step-1) Determine bucketLayout always using whether the parentID > 0, irrespective of bucket exists or not. 
```
        if (keysPerBucket.getKeysCount() > 0 &&
            keysPerBucket.getKeys(0).hasParentID()) {
          bucketLayout = BucketLayout.FILE_SYSTEM_OPTIMIZED;
        } else {
          bucketLayout = BucketLayout.DEFAULT;
        }
```

Reply:
According to HDDS-6556, I think we can just remove the bucket quota update logic here.
I have submitted HDDS-6563 to revert it, and I will rework this patch later.

Reply:
Sounds good. I really like the [HDDS-6563](https://issues.apache.org/jira/browse/HDDS-6563) idea.